### PR TITLE
net processing: Move CNode::nServices and CNode::nLocalServices to Peer

### PR DESCRIPTION
**Pull request by dergoegge**
Another step in #19398. Which services we offer to a peer and which services they offer to us is application layer data and should not be stored on `CNode`.

This is also a prerequisite for adding `PeerManager` unit tests (See #25515).

### CODE REVIEW

File: src/test/fuzz/util.cpp [hunk 1]
```diff
@@ -324,7 +324,6 @@ void FillNode(FuzzedDataProvider& fuzzed_data_provider, ConnmanTestMsg& connman,
     if (node.fDisconnect) return;
     assert(node.nVersion == version);
     assert(node.GetCommonVersion() == std::min(version, PROTOCOL_VERSION));
-    assert(node.nServices == remote_services);
```
Comment (MarcoFalke) [inline src/test/fuzz/util.cpp:327]: why is this removed when it can be moved down?

File: src/test/fuzz/util.h [hunk 2]
```diff
@@ -296,7 +296,6 @@ template <bool ReturnUniquePtr = false>
 auto ConsumeNode(FuzzedDataProvider& fuzzed_data_provider, const std::optional<NodeId>& node_id_in = std::nullopt) noexcept
 {
     const NodeId node_id = node_id_in.value_or(fuzzed_data_provider.ConsumeIntegralInRange<NodeId>(0, std::numeric_limits<NodeId>::max()));
-    const ServiceFlags local_services = ConsumeWeakEnum(fuzzed_data_provider, ALL_SERVICE_FLAGS);
```
Comment (MarcoFalke) [inline src/test/fuzz/util.h:299]: why is this removed and replaced with a hardcoded value?

File: src/net_processing.cpp [hunk 3]
```diff
@@ -636,7 +653,7 @@ class PeerManagerImpl final : public PeerManager
     /** Get a pointer to a mutable CNodeState. */
     CNodeState* State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
-    uint32_t GetFetchFlags(const CNode& pfrom) const EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    uint32_t GetFetchFlags(const Peer& peer) const EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
Comment (jnewbery) [inline src/net_processing.cpp:656]: This function no longer requires `cs_main`.

File: src/net_processing.cpp [hunk 4]
```diff
@@ -757,7 +774,7 @@ class PeerManagerImpl final : public PeerManager
     /** Update pindexLastCommonBlock and add not-in-flight missing successors to vBlocks, until it has
      *  at most count entries.
      */
-    void FindNextBlocksToDownload(NodeId nodeid, unsigned int count, std::vector<const CBlockIndex*>& vBlocks, NodeId& nodeStaller) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    void FindNextBlocksToDownload(NodeId nodeid, const Peer& peer, unsigned int count, std::vector<const CBlockIndex*>& vBlocks, NodeId& nodeStaller) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
Comment (jnewbery) [inline src/net_processing.cpp:777]: Perhaps it makes sense to remove `NodeId nodeid` as an argument to this function, now that it can easily be accessed from the `peer` argument.

File: src/rpc/net.cpp [hunk 5]
```diff
@@ -130,6 +125,11 @@ static RPCHelpMan getpeerinfo()
                     {RPCResult::Type::BOOL, "inbound", "Inbound (true) or Outbound (false)"},
                     {RPCResult::Type::BOOL, "bip152_hb_to", "Whether we selected peer as (compact blocks) high-bandwidth peer"},
                     {RPCResult::Type::BOOL, "bip152_hb_from", "Whether peer selected us as (compact blocks) high-bandwidth peer"},
+                    {RPCResult::Type::STR_HEX, "services", /*optional=*/true, "The services offered"},
```
Comment (jnewbery) [inline src/rpc/net.cpp:128]: I'm not sure if changing these fields to optional requires a release note, since missing fields that used to be always present could break clients that assume they'll be there. See #25176 for example.

File: src/net.cpp [hunk 6]
```diff
@@ -235,10 +235,10 @@ bool IsPeerAddrLocalGood(CNode *pnode)
 
 std::optional<CAddress> GetLocalAddrForPeer(CNode *pnode)
 {
-    CAddress addrLocal = GetLocalAddress(&pnode->addr, pnode->GetLocalServices());
+    CAddress addrLocal = GetLocalAddress(&pnode->addr);
     if (gArgs.GetBoolArg("-addrmantest", false)) {
         // use IPv4 loopback during addrmantest
-        addrLocal = CAddress(CService(LookupNumeric("127.0.0.1", GetListenPort())), pnode->GetLocalServices());
+        addrLocal = CAddress(CService(LookupNumeric("127.0.0.1", GetListenPort())), NODE_NONE);
```
Comment (jnewbery) [inline src/net.cpp:241]: I think the line lower down:

```
addrLocal = CAddress{pnode->GetAddrLocal(), addrLocal.nServices, addrLocal.nTime};
```

No longer makes sense, since `addrLocal.nServices` will always be `NODE_NONE`. Perhaps change `addrLocal.nServices` to `NODE_NONE`, since it is the caller's job to set the services.

Perhaps even better would be to change `GetLocalAddrForPeer` and `GetLocalAddress` to return a `CService` instead of a `CAddress`, so the caller has to set the `nTime` and `nServices` fields (which might help avoid bugs like #25314).

File: src/test/fuzz/util.h (hunk 2 above)
Comment (jnewbery) [reply src/test/fuzz/util.h:299]: Which hardcoded value? The `CNode` object no longer contains `nServices`, so there's nothing to set here.

**Review by jnewbery — COMMENTED** [review]
Concept ACK.

File: src/test/fuzz/util.h (hunk 2 above)
Comment (MarcoFalke) [reply src/test/fuzz/util.h:299]: The symbol is moved from `CNode` to `Peer`. I think a refactoring pull request should not change the value of the symbol.

If the local service flags were previously determined by the fuzz engine, they should still be after this change, regardless of the namespace of the symbol.

Comment (jnewbery) [reply src/test/fuzz/util.h:299]: To be clear, are you suggesting that the `InitializeNode()` calls in fuzz/processmessage.cpp and fuzz/processmessages.cpp should be updated to take `ConsumeWeakEnum(fuzzed_data_provider, ALL_SERVICE_FLAGS)`?

Comment (MarcoFalke) [reply src/test/fuzz/util.h:299]: Yes.

I am also thinking if `InitializeNode` can be moved into `FillNode`, so that only one place needs to be updated.

File: src/test/fuzz/util.cpp (hunk 1 above)
Comment (dergoegge) [reply src/test/fuzz/util.cpp:327]: Moved the assert down.

File: src/test/fuzz/util.h (hunk 2 above)
Comment (dergoegge) [reply src/test/fuzz/util.h:299]: Moved all `InitializeNode` calls (there were only 2) into `FillNode` with service flags consumed from the fuzzer.

File: src/net_processing.cpp (hunk 4 above)
Comment (dergoegge) [reply src/net_processing.cpp]: Yup, removed the NodeId arg.

File: src/net.cpp (hunk 6 above)
Comment (dergoegge) [reply src/net.cpp:239]: > Perhaps even better would be to change GetLocalAddrForPeer and GetLocalAddress to return a `CService`

Added an extra commit to do this. Makes it obvious that the caller is supposed to set the `nTime` and `nServices` fields.

File: src/rpc/net.cpp (hunk 5 above)
Comment (dergoegge) [reply src/rpc/net.cpp]: I changed this back to be non-optional just to stick with the previous behavior (i.e. when the peer has not sent a `version` message yet we default to  `NODE_NONE`).

Comment (jnewbery) [reply src/rpc/net.cpp]: :+1: Seems pragmatic

**Comment by jnewbery** [comment]
utACK d5b6ba717686ccb06ae06a553f2a563e97f53327

File: src/test/denialofservice_tests.cpp [hunk 7]
```diff
@@ -34,6 +36,56 @@ static CService ip(uint32_t i)
...
+    connman.ProcessMessagesOnce(node);
+
+    // Receive messages
+    {
+        LOCK(node.cs_sendProcessing);
+        peerman.SendMessages(&node);
+    }
+    BOOST_CHECK_EQUAL(node.fSuccessfullyConnected, true);
+
+    TestOnlyResetTimeData();
```
Comment (dergoegge) [inline src/test/denialofservice_tests.cpp:90]: The Ci was failing because the peer manager calls `AddTimeData`  when receiving a version message and since the time offset data for adjusted time is a global this can mess with other tests. Calling `TestOnlyResetTimeData()` here should fix it.

**Comment by jnewbery** [comment]
utACK c38a00d0e3

File: src/test/denialofservice_tests.cpp [hunk 8]
```diff
@@ -356,7 +353,7 @@ BOOST_AUTO_TEST_CASE(peer_discouragement)
                          ConnectionType::INBOUND,
                          /*inbound_onion=*/false};
     nodes[0]->SetCommonVersion(PROTOCOL_VERSION);
-    peerLogic->InitializeNode(nodes[0], NODE_NETWORK);
+    peerLogic->InitializeNode(nodes[0], ServiceFlags(NODE_NETWORK | NODE_WITNESS));
```
Comment (MarcoFalke) [inline src/test/denialofservice_tests.cpp:356]: d4759a02b51368870526f04558797bb7f5749c73: Looks like you are changing the value, why?

File: src/test/net_tests.cpp [hunk 9]
```diff
@@ -857,7 +857,7 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
         *static_cast<TestChainState*>(&m_node.chainman->ActiveChainstate());
     chainstate.JumpOutOfIbd();
 
-    m_node.peerman->InitializeNode(&peer);
+    m_node.peerman->InitializeNode(&peer, ServiceFlags(NODE_NETWORK|NODE_WITNESS));
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/test/net_tests.cpp:849]: If this is supposed to be an alias for `peer.nLocalServices`, why not simply type that to make review easier?

**Review by maflcko — COMMENTED** [review]
Concept ACK, but I'd prefer if behaviour changes (even if they are in tests) are either explained in the commit description or done in a separate commit from the refactoring commit.

File: src/test/denialofservice_tests.cpp [hunk 10]
```diff
@@ -63,10 +114,9 @@ BOOST_AUTO_TEST_CASE(outbound_slow_chain_eviction)
                      /*addrNameIn=*/"",
                      ConnectionType::OUTBOUND_FULL_RELAY,
                      /*inbound_onion=*/false};
-    dummyNode1.SetCommonVersion(PROTOCOL_VERSION);
 
-    peerLogic->InitializeNode(&dummyNode1, ServiceFlags(NODE_NETWORK|NODE_WITNESS));
-    dummyNode1.fSuccessfullyConnected = true;
+    peerman.InitializeNode(&dummyNode1, ServiceFlags(NODE_NETWORK|NODE_WITNESS));
+    ConnectNode(connman, peerman, dummyNode1, ServiceFlags(NODE_NETWORK|NODE_WITNESS));
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/test/denialofservice_tests.cpp:122]: This is done because this PR removes `CNode::fClient` and `CNode::m_limited_node` members. These two variables are default initialized as `false` and then set to the correct value when a `version` message is received. In production this is fine, as the two bools are not used if the version handshake hasn't happened. In the test however, we don't do the version handshake and manually set `fSuccessfullyConnected = true`. The default `false` then results in the logic behaving as expected (e.g. the node is not a client, so we ask them for headers).

Now with the PR the two bools are removed and replaced with static functions that operate on `Peer::their_services` to determine client and pruned node status. In the test here, if we keep `Peer::their_services` as `NODE_NONE` then we would consider `dummyNode1` a client and not ask for headers.

I think this makes the test more realistic (also mentioned in the commit message) and shouldn't change behavior.

File: src/test/net_tests.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/test/net_tests.cpp]: Changed this to pass `GetLocalServices()` to all the `InitializeNode` calls and only later on when `CNode::nLocalServices` is actually removed, do we set the service flags directly (corresponding to their previous values).

File: src/net_processing.cpp [hunk 11]
```diff
@@ -207,6 +207,27 @@ struct Peer {
...
+     *  fulfill this role after IBD completes. P2P code is written in such a
+     *  way that it can gracefully handle peers who don't make good on their
+     *  service advertisements.
+     *
+     *  TODO: remove redundant CNode::nLocalServices*/
+    const ServiceFlags m_our_services;
+    /** Services this peer offered to us.
+     *
+     * TODO: remove redundant CNode::nServices */
+    std::atomic<ServiceFlags> m_their_services{NODE_NONE};
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/net_processing.cpp]: Removed the `m_` prefix.

File: src/test/denialofservice_tests.cpp (hunk 8 above)
Comment (dergoegge) [reply src/test/denialofservice_tests.cpp]: All the service flags should now have the same values as before.

File: src/test/denialofservice_tests.cpp [hunk 12]
```diff
@@ -34,6 +36,57 @@ static CService ip(uint32_t i)
...
+
+    // Receive messages
+    {
+        LOCK(node.cs_sendProcessing);
+        peerman.SendMessages(&node);
+    }
+
+    TestOnlyResetTimeData();
+
+    if (node.fDisconnect) return;
```
Comment (MarcoFalke) [inline src/test/denialofservice_tests.cpp:70]: c132abf1667ea9367ce797d808d6b6c885196601: Why would this disconnect? Seems like copy-pasted dead code?

Also, it looks like you forgot to copy-paste the peer-state asserts.

Also, I don't see why this is copied in the first place. In the commit message you claim that it is better to make the version handshake consistent with "live code". So wouldn't it be better to make this helper easily accessible for all tests instead of forcing reviewers to re-do the copy-paste every time it is and will be used?

See https://github.com/bitcoin/bitcoin/pull/25591, which should also make this pull's diff smaller.

File: src/net.h [hunk 13]
```diff
@@ -234,7 +235,7 @@ void RemoveLocal(const CService& addr);
 bool SeenLocal(const CService& addr);
 bool IsLocal(const CService& addr);
 bool GetLocal(CService &addr, const CNetAddr *paddrPeer = nullptr);
-CAddress GetLocalAddress(const CNetAddr *paddrPeer, ServiceFlags nLocalServices);
+CAddress GetLocalAddress(const CNetAddr *paddrPeer);
```
Comment (MarcoFalke) [inline src/net.h]: 9d9a6677c483772fe636b61480dfdde7d4586073 (nit): Could run clang-format on this line while touching it?

**Review by maflcko — COMMENTED** [review]
Sorry for the incremental review, but I left more questions.

File: src/net_processing.cpp [hunk 14]
```diff
@@ -1299,15 +1321,15 @@ void PeerManagerImpl::UpdateLastBlockAnnounceTime(NodeId node, int64_t time_in_s
     if (state) state->m_last_block_announcement = time_in_seconds;
 }
 
-void PeerManagerImpl::InitializeNode(CNode *pnode)
+void PeerManagerImpl::InitializeNode(CNode *pnode, ServiceFlags our_services)
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/net_processing.cpp]: Done.

File: src/test/denialofservice_tests.cpp (hunk 12 above)
Comment (dergoegge) [reply src/test/denialofservice_tests.cpp]: Rebased on top of the changes in #25591. This test now uses `ConnmanTestMsg::Handshake()`.

File: src/net_processing.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/net_processing.cpp]: I'm going to push back against this. It's entirely possible that `Peer` picks up some member functions in future, and therefore becomes a class, in which case these data members would be named incorrectly and inconsistently.

File: src/net.h [hunk 15]
```diff
@@ -625,7 +604,7 @@ class NetEventsInterface
 {
 public:
     /** Initialize a peer (setup state, queue any initial messages) */
-    virtual void InitializeNode(CNode* pnode) = 0;
+    virtual void InitializeNode(CNode& pnode, ServiceFlags our_services) = 0;
```
Comment (jnewbery) [inline src/net.h:607]: This is beginning to feel very mission-creepy, but if you're changing the argument from being a pointer to a ref, then the name of the argument should also change from `pnode` to `node` (p is the prefix for pointer).

Same comment internally in net_processing for the `InitializeNode()` methods.

File: src/test/denialofservice_tests.cpp [hunk 16]
```diff
@@ -8,13 +8,15 @@
 #include <chainparams.h>
 #include <net.h>
 #include <net_processing.h>
+#include <netmessagemaker.h>
```
Comment (MarcoFalke) [inline src/test/denialofservice_tests.cpp:11]: nit in the second commit: Can be removed?

File: src/test/denialofservice_tests.cpp [hunk 17]
```diff
@@ -44,11 +46,10 @@ BOOST_FIXTURE_TEST_SUITE(denialofservice_tests, TestingSetup)
 // work.
 BOOST_AUTO_TEST_CASE(outbound_slow_chain_eviction)
 {
-    auto connman = std::make_unique<CConnman>(0x1337, 0x1337, *m_node.addrman, *m_node.netgroupman);
+    ConnmanTestMsg& connman = *static_cast<ConnmanTestMsg*>(m_node.connman.get());
```
Comment (MarcoFalke) [inline src/test/denialofservice_tests.cpp:49]: nit in the second commit: Can be written shorter as `static_cast<ConnmanTestMsg&>(*m_node.connman)`?

File: src/test/denialofservice_tests.cpp [hunk 18]
```diff
@@ -44,11 +46,10 @@ BOOST_FIXTURE_TEST_SUITE(denialofservice_tests, TestingSetup)
...
     // Disable inactivity checks for this test to avoid interference
-    static_cast<ConnmanTestMsg*>(connman.get())->SetPeerConnectTimeout(99999s);
-    auto peerLogic = PeerManager::make(*connman, *m_node.addrman, nullptr,
-                                       *m_node.chainman, *m_node.mempool, false);
+    connman.SetPeerConnectTimeout(99999s);
+    PeerManager& peerman = *m_node.peerman.get();
```
Comment (MarcoFalke) [inline src/test/denialofservice_tests.cpp:52]: Same: no need for get?

File: src/test/denialofservice_tests.cpp [hunk 19]
```diff
@@ -63,10 +64,16 @@ BOOST_AUTO_TEST_CASE(outbound_slow_chain_eviction)
...
                      ConnectionType::OUTBOUND_FULL_RELAY,
                      /*inbound_onion=*/false};
-    dummyNode1.SetCommonVersion(PROTOCOL_VERSION);
 
-    peerLogic->InitializeNode(dummyNode1, dummyNode1.GetLocalServices());
-    dummyNode1.fSuccessfullyConnected = true;
+    connman.Handshake(
+        /*node=*/dummyNode1,
+        /*successfully_connected=*/true,
+        /*remote_services=*/ServiceFlags(NODE_NETWORK | NODE_WITNESS),
```
Comment (MarcoFalke) [inline src/test/denialofservice_tests.cpp:70]: second commit: Maybe mention in the commit message that this is a behaviour change that is needed for the next commit? Previously this was NODE_NODE, now it is something else?

File: src/net.h [hunk 20]
```diff
@@ -376,6 +374,8 @@ class CNode
 
     uint64_t nRecvBytes GUARDED_BY(cs_vRecv){0};
 
+    /** Whether this peer provides all services that we want. Used for eviction decisions */
+    std::atomic_bool m_has_all_wanted_services{false};
     std::atomic<std::chrono::seconds> m_last_send{0s};
```
Comment (MarcoFalke) [inline src/net.h:379]: nit in b41c44dcdc49941c4c2e70bcdfdcd8b1d956c69f: Maybe move `m_has_all_wanted_services` down to the other eviction stuff (m_last_tx_time, ...). Otherwise, the comment "Used for eviction decisions" will (incorrectly) imply that m_last_send is also used for eviction, when it is currently not.

**Review by maflcko — APPROVED** [review]
Looking a lot better with the Handshake test helper. Thanks.

review ACK 8f6ed20acae2a4f93f5b52c3937ab704980db44b though I have not yet reviewed the last two commits 🏼

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 8f6ed20acae2a4f93f5b52c3937ab704980db44b though I have not yet reviewed the last two commits 🏼
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjqHwwAlNYcnHxFpa3FBWcBsYy2DSbbm3Ww6p+dqZLivANiuNWpc7qKrFYoVQKY
OlCbOyMJNHuGbC0ACwjSfVxh1VhyLpdW7qtrMr203XD3j3MRComWe1Q5h6jG4jfn
rEB0ybHIAowvoWQBfvtY5Ti20WEsiPvWL4qEgA+b+wYClG+42BCHUyRePkN+1k+u
0fD/BOhTiRszyXgfWNV5q1iI/UhR6caf213mV/5X+udIXWMgi1PRV+4HQG4GmZdq
s25Ua9Quzmgtjz3sUDb4op2ztlS8BOMpzulKqSiXmIyAQz72zODaYrCieHOItZs8
vRR/JbGs4IMF/whQFtzJgPIMQ+O6iO8V/VHBoYbrmhjcbxpX9ioXhpX4mXelepeN
NFkdOR9WQWpaMAfcOl948axSgfVaN7MFU+9+0rzpHLEamQj5wbzdP4xdtyXqFGA4
jf11QGqoAh3a0IvYhEoUgK0MPbeR6ASSvhzAtglIw6vFtCPkCXBRnM+8vqBmaLBr
iVM6GTFy
=YLIt
-----END PGP SIGNATURE-----
```


</details>

File: src/qt/rpcconsole.cpp [hunk 21]
```diff
@@ -1197,6 +1196,7 @@ void RPCConsole::updateDetailWidget()
     // This check fails for example if the lock was busy and
     // nodeStateStats couldn't be fetched.
     if (stats->fNodeStateStatsAvailable) {
+        ui->peerServices->setText(GUIUtil::formatServicesStr(stats->nodeStateStats.their_services));
```
Comment (MarcoFalke) [inline src/qt/rpcconsole.cpp:1199]: Tested the result when the stats aren't available, which looks good:

Screenshot from 2022-07-13 14-07-22

**Review by maflcko — COMMENTED** [review]
I tried to do some testing, but the second commit doesn't compile:

```
test/denialofservice_tests.cpp:75:23: error: too many arguments to function call, expected 6, have 7

File: src/test/denialofservice_tests.cpp (hunk 16 above)
Comment (dergoegge) [reply src/test/denialofservice_tests.cpp]: Done.

File: src/test/denialofservice_tests.cpp (hunk 17 above)
Comment (dergoegge) [reply src/test/denialofservice_tests.cpp]: Yes, done.

File: src/test/denialofservice_tests.cpp (hunk 19 above)
Comment (dergoegge) [reply src/test/denialofservice_tests.cpp:69]: It is actually also needed in this commit because `CNode::fClient` and `CNode::m_limited_node` are set during the version handshake. The commit message now mentions this.

File: src/net.h (hunk 20 above)
Comment (dergoegge) [reply src/net.h]: Done.

File: src/net.cpp [hunk 22]
```diff
@@ -236,10 +236,10 @@ bool IsPeerAddrLocalGood(CNode *pnode)
...
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/net.cpp]: Switched the order of the commits. `NODE_NONE` is not used anymore now.

File: src/net.h (hunk 15 above)
Comment (dergoegge) [reply src/net.h]: Done.

File: src/net.cpp [hunk 23]
```diff
@@ -206,15 +206,13 @@ static std::vector<CAddress> ConvertSeeds(const std::vector<uint8_t> &vSeedsIn)
 // Otherwise, return the unroutable 0.0.0.0 but filled in with
 // the normal parameters, since the IP may be changed to a useful
 // one by discovery.
-CAddress GetLocalAddress(const CNetAddr *paddrPeer, ServiceFlags nLocalServices)
+CService GetLocalAddress(const CNetAddr *paddrPeer)
```
Comment (MarcoFalke) [inline src/net.cpp:209]: nit in 4c59e37c18b177fc50786de37ffffb52de6daaa7: Only if you retouch, clang-format this line

File: src/net.cpp [hunk 24]
```diff
@@ -233,12 +231,12 @@ bool IsPeerAddrLocalGood(CNode *pnode)
            IsReachable(addrLocal.GetNetwork());
 }
 
-std::optional<CAddress> GetLocalAddrForPeer(CNode *pnode)
+std::optional<CService> GetLocalAddrForPeer(CNode *pnode)
```
Comment (MarcoFalke) [inline src/net.cpp:234]: same

File: src/net.cpp [hunk 25]
```diff
@@ -262,7 +260,7 @@ std::optional<CAddress> GetLocalAddrForPeer(CNode *pnode)
     if (addrLocal.IsRoutable() || gArgs.GetBoolArg("-addrmantest", false))
     {
         LogPrint(BCLog::NET, "Advertising address %s to peer=%d\n", addrLocal.ToString(), pnode->GetId());
-        return addrLocal;
+        return {addrLocal};
```
Comment (MarcoFalke) [inline src/net.cpp:263]: 4c59e37c18b177fc50786de37ffffb52de6daaa7: Is this change needed? For me it compiles without.

File: src/net.h [hunk 26]
```diff
@@ -144,8 +144,8 @@ enum
 };
 
 bool IsPeerAddrLocalGood(CNode *pnode);
-/** Returns a local address that we should advertise to this peer */
-std::optional<CAddress> GetLocalAddrForPeer(CNode *pnode);
+/** Returns a local address that we should advertise to this peer. */
+std::optional<CService> GetLocalAddrForPeer(CNode* pnode);
```
Comment (MarcoFalke) [inline src/net.h:148]: 4c59e37c18b177fc50786de37ffffb52de6daaa7: While touching this method, could switch `*` to `&`? (And remove "p")

File: src/net.h [hunk 27]
```diff
@@ -163,7 +163,7 @@ void RemoveLocal(const CService& addr);
 bool SeenLocal(const CService& addr);
 bool IsLocal(const CService& addr);
 bool GetLocal(CService &addr, const CNetAddr *paddrPeer = nullptr);
-CAddress GetLocalAddress(const CNetAddr *paddrPeer, ServiceFlags nLocalServices);
+CService GetLocalAddress(const CNetAddr* paddrPeer);
```
Comment (MarcoFalke) [inline src/net.h:166]: Same

**Review by maflcko — APPROVED** [review]
lgtm

review ACK e7aea6ae6d0cd668dc7228cc24d0a58419c6ebcd 🏞

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK e7aea6ae6d0cd668dc7228cc24d0a58419c6ebcd 🏞
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjuqQv/aHcMUNox1MQBi+1Mj3WJorgEHy85MaRS3JiylOIrpUpGgtjvmh91HpoG
6ThPApvoEKGf3sYRKgU8COOVoEgzBj3AqXeG/RTu6wJaZNq3AFRmCjBtzzsxJjDJ
6w9SLoBexTQyTLLFEf/EvwxlE3ttldDgPQOQ0Ui6zZgxs1eCjpgBb119V41BTBmy
esOxtkE8ZHoPo+4nPYhB0d9f2i60wYX1jbeZAv0b3BDB+4XL67UUXhq8IhEJFvRY
LTDbgtSxhNQJAx3NOcOeHtmDdy8xfQe3Nnc7q8WAIfcitCa+DGXSIt5pKTzuwvut
bbPZroBqWarME+zSsVxvR6JMfAVoTrLO2zaFoAYrCvFQq6yrbq3OJZ6Rqq9GOJjj
kuwMGpW6yKJvTRb0qxgcds3Bgh6UJ/U3ZUVtfxNJ5LR7+JCpOkSBN1Ro+vH+amml
GZC7tDdeC7f8IaexX51tUXbweRzE49zK3NH5pYazHubyZDncMRM4dlLs8AtIzL8C
6XsRJRHp
=ZDto
-----END PGP SIGNATURE-----
```


</details>

**Comment by jnewbery** [comment]
@dergoegge just checking that you saw my comment about member naming: https://github.com/bitcoin/bitcoin/pull/25514#discussion_r919936826.

File: src/net_processing.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/net_processing.cpp:225]: That seems reasonable, re-added the `m_` prefix.

**Comment by MarcoFalke** [comment]
ACK 8d8eeb422e64ccd08bce92eed2ee9cbbf55ffd67 🔑

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 8d8eeb422e64ccd08bce92eed2ee9cbbf55ffd67 🔑
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjJZQv/TaiI8pAKqRBeggxNkZdPjDY4gnn1vJgyfjjDsVu1FrKZBbOm0AxGPmSP
o3N2x5ZZWcjIzU6Rmv0aWqxmz1yaHUByF9bHC8PBcfcdnb9ELVePbYApM3pcFZU1
DX/wR1RJcb+m9/h8Kl8f6wQMIHf/2fCuqi10/AX+RqLIgEpIRECnEGaHVE6jGV2k
gI+KDnQRxRBelpSloDK7lnTAOUbU0H9ReeILtpIDM6HNGynC50C2VEj29cSxjBdD
KaqNPbbcHyqXiDM0DhNtFRRWNS8PP2d46xWOK+yWr/sInPpmdns8xsA29s3FwUp0
mqpseRqrfRaFOZRqtDjVX1nxs3JeQNSkxCnJoVFMReWzZcvEe9PO0jZjOrntCAwU
cxJcrNsRlaxkIT3Ty2mKYuoRZKujxXUmHzBD4KOYiLd0DNHX+xCEcEb/aQFQM21g
cLOkVzMe8IDfocPtbKw38HQSHgsh/ppzWM61KBp8vBuI4jTlGY3XUieUFW6mlhgj
vGhOSGp+
=CVY2
-----END PGP SIGNATURE-----
```


</details>

File: src/net.h [hunk 28]
```diff
@@ -1035,16 +1015,14 @@ class CConnman
...
-     * connection (in ConnectNode()) under a member also called
-     * nLocalServices.
+     * This data is replicated in each Peer instance we create.
      *
      * This data is not marked const, but after being set it should not
-     * change. See the note in CNode::nLocalServices documentation.
+     * change.
      *
-     * \sa CNode::nLocalServices
+     * \sa Peer::our_services
```
Comment (jnewbery) [inline src/net.h:1025]:
```suggestion
     * \sa Peer::m_our_services
```

**Review by jnewbery — APPROVED** [review]
utACK 8d8eeb422e64ccd08bce92eed2ee9cbbf55ffd67

One missed `our_services` -> `m_our_services` in a comment, but that shouldn't hold up merge.

**Review by mzumsande — COMMENTED** [review]
Code Review ACK 8d8eeb422e64ccd08bce92eed2ee9cbbf55ffd67

This looks good to me, only found nits/doc comments during review.

**Comment by dhruv** [comment]
@dergoegge @jnewbery I am still digging deeper here, but I have a question (and might be terribly wrong). Moving `nServices` and `nLocalServices` into `Peer` seems to imply that the service bits only gain importance during the VERSION/VERACK handshake only useful in the P2P layer, not at the transport layer.

However, I think service bits (also propagated with addr relay) are useful at the transport layer. For instance in BIP324, the [proposed signaling method](https://gist.github.com/dhruv/5b1275751bc98f3b64bcafce7876b489#signaling-v2-support) uses `nServices` in addr relay to branch logic in the establishment of the transport. Moving the service flags to `net_processing` will mean that I'll end up having to pass a boolean around. Is that an intended effect?

**Comment by dergoegge** [comment]
@dhruv Currently all service bits signal readiness to serve application layer data (blocks, bloom/compact filters, witness data). The BIP324 service bit differs from that as it signals support for a new transport protocol, which would make it the only bit that is needed within our net layer. I would therefore argue that passing around a bool to/within the net layer for the v2 bit seems more appropriate than passing around all service bits.
